### PR TITLE
rfcs: Add RFC for public SDPA primitive

### PR DESCRIPTION
**Pull request by saniljoshi**
# Description
This RFC proposes promoting Scaled Dot-Product Attention (SDPA) to a
first-class public primitive in oneDNN through a new public `dnnl::sdpa`
C++ API backed by the existing internal SDPA primitive plumbing. The first
implementation PR's scope is a CPU reference path for forward inference,
attention-mask support, causal-mask support, stride-aware layout handling,
and benchdnn validation. No optimized SDPA kernel is introduced in the
initial PR.

**Motivation / context.** oneDNN currently supports attention through the Graph
API and has internal SDPA primitive infrastructure, but it does not expose a
public SDPA primitive that framework consumers can call directly. A public
`dnnl::sdpa` primitive gives frameworks such as PyTorch, TensorFlow, ONNX
Runtime, and vLLM a stable primitive API entry point for fused attention,
while allowing optimized CPU implementations to be registered behind the same
API in follow-up work. ZenDNN zentorch validation shows the value of this path
on representative Torch Inductor dashboard models.

This PR adds only the RFC document (`rfcs/20260618-sdpa-primitive/`);
it contains no library code. Implementation lands in follow-up PRs per the
scope and acceptance criteria described in the RFC.

Rendered RFC:
https://github.com/amd/oneDNN-Zen/tree/zen-rfcs/rfcs/20260618-sdpa-primitive

POC branch
https://github.com/amd/oneDNN-Zen/tree/zendnn-onednn-sdpa

Fixes # N/A — design document (RFC), no associated bug.

Fixes # (github issue)

# Checklist

## General

- [ ] Do all unit and benchdnn tests (`make test` and `make test_benchdnn_*`) pass locally for each commit?
- [ ] Have you formatted the code using clang-format?

## Performance improvements

- [ ] Have you submitted performance data that demonstrates performance improvements?

### New features

- [x] Have you published an RFC for the new feature?
- [ ] Was the RFC approved?
- [ ] Have you added relevant tests?

### Bug fixes

- [ ] Have you included information on how to reproduce the issue (either in a github issue or in this PR)?
- [ ] Have you added relevant regression tests?

## [RFC](https://github.com/uxlfoundation/oneDNN/tree/rfcs) PR

- [x] Does RFC document follow the [template](https://github.com/uxlfoundation/oneDNN/blob/rfcs/rfcs/template.md#onednn-design-document-rfc)?
- [x] Have you added a link to the rendered document?

cc:- @avinashcpandey, @amukho

### CODE REVIEW

**Comment by amukho** [comment]
hi @vpirogov it would be great if you can review this and also add other relevant reviewers

**Comment by vpirogov** [comment]
> hi @vpirogov it would be great if you can review this and also add other relevant reviewers

In current form this RFC is hard to review because it's unnecessarily verbose and repetitive. I would really appreciate if it is reworked to clearly and concisely formulate the problem statement and proposed solution using guidelines in [RFC template](https://github.com/uxlfoundation/oneDNN/blob/rfcs/rfcs/template.md).

File: rfcs/20260618-sdpa-primitive/README.md [hunk 1]
```diff
@@ -0,0 +1,458 @@
...
+   oneDNN already contains the SDPA primitive descriptor and GPU kernels
+   (`src/common/sdpa_*`, `sdpa_primitive_desc_create`), yet the only way to
+   reach it is a **test-only** wrapper in
+   `tests/gtests/internals/sdpa_internal.hpp`.
+   The primitive plumbing exists but was parked behind the test/internal
+   interface rather than promoted to a supported `dnnl::sdpa` API, and its CPU
+   path is effectively absent.
+
+In short: the supported path (Graph) is not a primitive, and the primitive that
+exists is not public.
```
Comment (TaoLv) [inline rfcs/20260618-sdpa-primitive/README.md]: I feel the question is why we need a public primitive for this?

File: rfcs/20260618-sdpa-primitive/README.md [hunk 2]
```diff
@@ -0,0 +1,458 @@
...
+## 1. Motivation
+
+Attention is the dominant compute pattern in transformer inference. Frameworks
+(PyTorch `scaled_dot_product_attention`, vLLM, ONNX Runtime) want a single,
+fused, well-optimised attention call. oneDNN is the CPU primitive library those
+frameworks depend on, yet it does **not** expose a public attention primitive.
+
+**Problem.** A consumer that wants oneDNN to run attention today must either:
+- build the **Graph** (assemble the MHA fusion pattern), which is heavier to
+  integrate and is graph-API-centric; or
```
Comment (TaoLv) [inline rfcs/20260618-sdpa-primitive/README.md]: This is not a problem. PyTorch already has native integration for SDPA with Graph API: https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/native/mkldnn/xpu/detail/Attention.cpp . Though currently the path is enabled for xpu only, the integration code and methodology can be reused for cpu as well.

Comment (saniljoshi) [reply rfcs/20260618-sdpa-primitive/README.md]: Yes, you are right that the existing PyTorch Graph API integration path can likely be reused for CPU as well.
My point is slightly different: when a framework already has SDPA as a fused op, it would be useful to have a direct oneDNN primitive API for that op. Otherwise the framework has to represent SDPA as individual operations and rely on Graph pattern matching to map it back to SDPA. A primitive API would provide a simpler direct path while still keeping the existing Graph API path valid.

**Comment by saniljoshi** [comment]
> > hi @vpirogov it would be great if you can review this and also add other relevant reviewers

Thanks @vpirogov for your inputs. We are working on to refine/reduce the verbosity of RFC. We will update the RFC by tomorrow.

File: rfcs/20260618-sdpa-primitive/README.md (hunk 2 above)
Comment (dzarukin) [reply rfcs/20260618-sdpa-primitive/README.md]: > ... it would be useful to have a direct oneDNN primitive API for that op.

> A primitive API would provide a simpler direct path...

These two statements are subjective and defined by a matter of taste.

In a case of a new feature on PyTorch API (e.g. a new flag or existing flag gets a new flavor that can't be expressed on oneDNN side with existing arguments) to support it in oneDNN, an API extension must happen.

In case of Graph API in worst case scenario it would be a new Op on the user surface level and pattern matching adjustment under the hood to support this new Op. Op is a stable abstraction in Graph. Stable means it doesn't change its core and can receive modifications through other mean like attributes. There are no examples of Op require rebuilding through the Graph history.

In case of a primitive, a new C API must be created since this is a foundation of stability for primitives. There will be two (or more) C APIs for SDPA primitive since that moment. Both must exist since the library provides full ABI and API compatibility between minor releases. The more features appear, the more new APIs calls must appear, all must be tested and supported. This is the situation that we push back.

So unless there's a strong technical impediment from the customer side that pushes the library to provide a primitive, this is an undesired path for the library from the maintenance perspective.

Hope it helps. Thanks.

**Comment by saniljoshi** [comment]
> > > hi @vpirogov it would be great if you can review this and also add other relevant reviewers
> 
> Thanks @vpirogov for your inputs. We are working on to refine/reduce the verbosity of RFC. We will update the RFC by tomorrow.

Hi @vpirogov ,
We have reworked the RFC to reduce duplicated content and have updated and refined the Motivation section to address the comments and questions raised by TaoLv and Zarukin. 


cc:  @avinashcpandey1

File: rfcs/20260618-sdpa-primitive/README.md (hunk 2 above)
Comment (saniljoshi) [reply rfcs/20260618-sdpa-primitive/README.md]: > > ... it would be useful to have a direct oneDNN primitive API for that op.
> 
> > A primitive API would provide a simpler direct path...
> 
> These two statements are subjective and defined by a matter of taste.
> 
> In a case of a new feature on PyTorch API (e.g. a new flag or existing flag gets a new flavor that can't be expressed on oneDNN side with existing arguments) to support it in oneDNN, an API extension must happen.
> 
> In case of Graph API in worst case scenario it would be a new Op on the user surface level and pattern matching adjustment under the hood to support this new Op. Op is a stable abstraction in Graph. Stable means it doesn't change its core and can receive modifications through other mean like attributes. There are no examples of Op require rebuilding through the Graph history.
> 
> In case of a primitive, a new C API must be created since this is a foundation of stability for primitives. There will be two (or more) C APIs for SDPA primitive since that moment. Both must exist since the library provides full ABI and API compatibility between minor releases. The more features appear, the more new APIs calls must appear, all must be tested and supported. This is the situation that we push back.
> 
> So unless there's a strong technical impediment from the customer side that pushes the library to provide a primitive, this is an undesired path for the library from the maintenance perspective.
> 
> Hope it helps. Thanks.


Thanks @dzarukin , this helps.

One clarification: the current proposal is not to add a new SDPA C API. The SDPA primitive descriptor plumbing and C entry point already exist in oneDNN today (`sdpa_primitive_desc_create`), but they are currently reachable only through the internal/test path (`sdpa_test_iface.hpp`).
What we are proposing is to reuse that existing primitive plumbing, move the entry point out of the test-only path into the supported implementation path, add the missing CPU reference implementation, and expose a C++ `dnnl::sdpa` wrapper on top. Since the existing descriptor creation path already carries the required descriptors, mask type, scale behavior, softmax mode, and attributes, we do not expect to need additional C API variants for the initial scope.

Also, we have updated the motivation section based on your comments.

Comment (dzarukin) [reply rfcs/20260618-sdpa-primitive/README.md]: It might be a surprising finding but _"to reuse that existing primitive plumbing, move the entry point out of the test-only path into the supported implementation path"_ identically equal to _"add a new SDPA C API"_.

This interface is not exposed at the public level for exact reasons explained in the post above.

We didn't expect it to modify much too, yet that interface had been updated 7 times since the introduction. And it won't surprise me if the industry will push it to the next limit in short time in the future now.

**Comment by vpirogov** [comment]
+@eikanwang for framework enabling perspective.

**Comment by vpirogov** [comment]
> We have reworked the RFC to reduce duplicated content and have updated and refined the Motivation section to address the comments and questions raised by TaoLv and Zarukin.

I really appreciate the effort, @saniljoshi! The style is dramatically better now.

File: rfcs/20260618-sdpa-primitive/README.md (hunk 2 above)
Comment (saniljoshi) [reply rfcs/20260618-sdpa-primitive/README.md]: @dzarukin  hope you are doing well.

I have one clarification question about the intended boundary here.

I understand the concern about exposing SDPA as a new public C/C++ primitive API. What I am trying to separate from that is the question of adding a CPU implementation behind the existing internal `primitive_kind::sdpa` path.

Today GPU has SDPA primitive implementations registered, and the standalone `benchdnn --sdpa` driver tests that internal path through `sdpa_primitive_desc_create()`. PR #4884 also describes this as internal SDPA primitive functionality, not public API. On CPU, the implementation list for `primitive_kind::sdpa` currently returns empty and the SDPA benchdnn driver skips CPU.

So is the concern specifically about exposing SDPA as a public primitive API, or does it also apply to adding CPU support behind the existing internal SDPA primitive path for parity and validation with GPU?

In short, I am trying to understand why the internal SDPA primitive path is validated directly for GPU, but not for CPU. Is that because CPU SDPA is expected to be supported only through Graph API, or is CPU support behind the existing internal SDPA primitive path still a valid direction?

cc: @avinashcpandey1, @vpirogov , @amukho

Comment (vpirogov) [reply rfcs/20260618-sdpa-primitive/README.md]: > What I am trying to separate from that is the question of adding a CPU implementation behind the existing internal primitive_kind::sdpa path.

Adding an optimized version (preferably based on ukernel API) to internal SDPA primitive would be a reasonable way to optimize SDPA in Graph API.

> Today GPU has SDPA primitive implementations registered, and the standalone benchdnn --sdpa driver tests that internal path through sdpa_primitive_desc_create().

SDPA functionality on GPU is primarily validated using benchdnn `--graph` driver and subgraphs captured from PyTorch and Tensorflow. Unlinke `--sdpa` driver this approach covers other implementations available under Graph API including `larger_partition_kernel_t` and `sdp_decomp_kernel_t`.

> In short, I am trying to understand why the internal SDPA primitive path is validated directly for GPU, but not for CPU. 

Internal SDPA primitive is not implemented for CPU, so there's nothing to validate. Currently Graph API dispatches `sdp_decomp_kernel_t` as an optimized implementation for CPU, which can be tested using benchdnn `--graph` driver.

Hope this helps!

+@TaoLv
